### DISCUSSION TITLE
Lightness inversion preserving colors

### DISCUSSION THREAD

**Discussion by mara004** [python-pillow/Pillow#7314] [Q&A] [answered]
Hello,

I'm wondering how to invert lightness of an image while preserving colors, using Pillow?
e.g. white to black (and vice versa), dark green to light green, etc.
I am not looking for classical color inversion (which would, for example, turn dark green into light pink).

A demo of what I mean is Okular's "Invert Lightness" accessibility option.
For some context, I would like to implement something similar (a dark mode for PDFs), but with an option to exclude images from the inversion, which is relevant for machine-generated (not scanned) PDFs.

Thanks!

**Comment by Yay295** [comment]
I think you could convert the image to HSV and then [invert](https://pillow.readthedocs.io/en/stable/reference/ImageChops.html#PIL.ImageChops.invert) the V channel.

**Reply by mara004** [reply]
Thanks for the suggestion, but this is not quite what I meant yet.
Inverting the V channel seems to keep colors as-is, but I still want to invert their lightness (e.g. dark blue to light blue).
"preserving colors" was probably a bit misleading here, sorry.

**Reply by mara004** [reply]
Default | Invert Colors | Invert Lightness | Invert V Channel
---|---|---|---
![image](https://github.com/python-pillow/Pillow/assets/65915611/747766d9-c052-474b-90f9-2086d5349f51) | ![image](https://github.com/python-pillow/Pillow/assets/65915611/f2a0dca8-5c73-4111-8a5e-462222fbb074) | ![image](https://github.com/python-pillow/Pillow/assets/65915611/ba2ec995-fea4-469e-8109-51c2a843b69b) | ![CinelerraGG_Manual_003](https://github.com/python-pillow/Pillow/assets/65915611/04be5ec3-dd10-4c83-8110-be857825dede)

**Reply by mara004** [reply]
I did some investigation: Okular's [`invertLightness()`](https://github.com/KDE/okular/blob/7e3497bb54dd5db3c8bb8267a0f5296a7a1aeaab/gui/pagepainter.cpp#L748) function (introduced with https://github.com/KDE/okular/commit/aa6833483e9a87b151e38dd82b695ea9f5ce7b8d) apparently uses HSL rather than HSV. However, I think HSL is not supported by Pillow yet?

**Reply by radarhere** [reply]
HSL isn't supported as a mode, no - https://pillow.readthedocs.io/en/stable/handbook/concepts.html#modes

You could use [`colorsys.rgb_to_hls`](https://docs.python.org/3/library/colorsys.html#colorsys.rgb_to_hls) and [`colorsys.hls_to_rgb`](https://docs.python.org/3/library/colorsys.html#colorsys.hls_to_rgb) to convert the values, then invert the lightness, then convert them back.

**Comment by homm** [comment, marked as answer]
Here is the full example using HLS color space:

```python
import colorsys
from PIL import Image, ImageFilter


def invert_colors(r, g, b):
    h, l, s = colorsys.rgb_to_hls(r, g, b)
    l = 1.0 - l
    return colorsys.hls_to_rgb(h, l, s)


invert_filter = ImageFilter.Color3DLUT.generate(17, invert_colors)
Image.open('257923775.png').filter(invert_filter).save('257923775.out.png')
```

![257923775 out](https://github.com/python-pillow/Pillow/assets/128982/ff9aa9ef-6687-4eb7-93f1-c15cd884c2fc)

**Reply by mara004** [reply]
Thanks @homm and @radarhere for the advice!

I had found how to do it with OpenCV yesterday (and there's also a colour-science package that might work), but PIL and a built-in module of course fits better with my current dependencies.

**Reply by mara004** [reply]
@homm I'm afraid I don't yet understand the `size` parameter of `ImageFilter.Color3DLUT.generate` (the docs didn't help me much). How do you get to the value `17`? Could you please explain?

**Reply by homm** [reply]
It’s a size of generated [3D lookup table](https://en.m.wikipedia.org/wiki/3D_lookup_table). The bigger the size, the close interpolated values to the real ones. 17 is a safe minimal, less could be noticeable on some transformations. In your case I don’t think there would be any differences between 17 and higher sizes.

**Reply by mara004** [reply]
Thanks!

**Comment by raboehm** [comment]
```python
def invlight(r, g, b):
    # lightness = (max(r,g,b) + min(r,g,b)) / 2
    # lightnessnew = 255 - lightness
    # deltalightness = lightnessnew - lightness = 255 - 2 * lightness
    dl = 255 - (max(r,g,b) + min(r,g,b))
    return [r + dl, g + dl, b + dl]
```
